### ISSUE TITLE
proposal normalize()  function on streams

### ISSUE THREAD

**Issue by aanthony1243** [influxdata/flux#1298]
the normalize function will scan all table schemas and apply operations to create a schema where each table in the stream may be deterministically mapped to the new schema (though some column values may be null).  

example:   
[_measurement:string, _field:string, host:string, cpu:float]
[_measurement:string, _field:string, host:string, region:string, disk:int]

resulting schema: 
[_measurement:string, _field:string, host:string, region:string, cpu:float, disk:int]

Both tables above will then be copied into new tables with the unified schema.  What are some unanswered questions?

1.  What would the group key be?  it should also be unified.  
2.  What if 2 columns have the same name but different types?

**Comment by chnn** [comment]
We ended up doing something very similar in [giraffe](https://github.com/influxdata/giraffe/) to interop with Flux data.

Basically, we wanted the public visualization API to be "pick columns/variables from a single table and map them to aesthetics of a plot" (i.e. we want the input data to be [tidy](https://vita.had.co.nz/papers/tidy-data.pdf)). We also made the simplifying assumption that for the purposes of a visualization, null values in a table are irrelevant—we simply do not plot them.

So when we [parse a Flux CSV](https://github.com/influxdata/giraffe/blob/a2b6985bbb9158ad41528d775025eac25a58228f/src/utils/fromFlux.ts#L25-L74), we spread every table into a single wide and sparse table (like the `normalize` function):

```ts
/*
  Convert a Flux CSV response to a `Table`.
  
  For example, given a series of Flux tables that look like this:
  
      column_a | column_b | column_c  <-- name
      long     | string   | long      <-- type
      ------------------------------
             1 |      "g" |       34
             2 |      "f" |       58 
             3 |      "c" |       21
  
      column_b | column_d   <-- name
      double   | boolean    <-- type
      -------------------
           1.0 |     true
           2.0 |     true
           3.0 |     true
  
  This function will spread them out to a single wide table that looks like
  this instead:
  
      column_a | column_b (string) | column_c | column_b (number) | column_d  <-- key
      column_a | column_b          | column_c | column_b          | column_d  <-- name
      number   | string            | number   | number            | bool      <-- type
      ---------------------------------------------------------------------
             1 |               "g" |       34 |                   | 
             2 |               "f" |       58 |                   |
             3 |                   |       21 |                   |
               |                   |          |               1.0 |     true
               |                   |          |               2.0 |     true
               |                   |          |               3.0 |     true
  
  
  ...snip...
  
  The `Table` stores a `key` for each column which is separate from the column
  `name`. If multiple Flux tables have the same column but with different
  types, they will be distinguished by different keys in the resulting `Table`;
  otherwise the `key` and `name` for each column in the result table will be
  identical.
*/
export const fromFlux = (fluxCSV: string): FromFluxResult => {
  // ...
}
``` 

> What if 2 columns have the same name but different types?

We give each column a "key" separate from its name. I think of the name as a convenient shorthand for the key, since most of the time they are the same.

> What would the group key be? it should also be unified.

We take the union of all group keys. I think this works well, since adding a column `k` to to a table's group key when that table doesn't contain `k` means that `k` is `null` for every row in the table. So the meaning of the group key doesn't change—it just has some extra noise.


Overall I am pretty happy with this approach, but I wish it was more of standard or "blessed" Flux concept.

Something I've thought about in the past: what if the wide/sparse format was a "dialect" or similar in the request? This would disable the streaming semantics of Flux (which we don't use) and send all the data once it is ready, in a normalized format.

I think being able to opt-in to request/response semantics instead of streaming semantics could enable some other nice features, like a JSON response format (much easier to work with), a `Content-Length` response header, and a single location to check for errors (rather than a table with an `error` column streamed at some point in the request).

**Comment by russorat** [comment]
Is this the same as "normalizing" two datasets to have the same range? For example:

> For example, consider a data set containing two features, age, and income(x2). Where age ranges from 0–100, while income ranges from 0–100,000 and higher. Income is about 1,000 times larger than age. So, these two features are in very different ranges. When we do further analysis, like multivariate linear regression, for example, the attributed income will intrinsically influence the result more due to its larger value. But this doesn’t necessarily mean it is more important as a predictor. So we normalize the data to bring all the variables to the same range.

**Comment by adrian-thurston** [comment]
I believe the proposed `normalize` is a simplified `pivot`. Or put another way, `pivot` is a generic "normalization" of *schemas*. It is not related to normalizing the values in the data, as most data scientists would come to understand "normalize".

Maybe @aanthony1243 can comment further, but I think given the existence of `pivot` we do not need the above `normalize` function.

**Comment by russorat** [comment]
thanks @adrian-thurston! maybe i should open a new issue to request the data scientist version of normalize?

**Comment by adrian-thurston** [comment]
Mathematically, it could just be about the use of existing functions to compute some normalized version of the data, as dictated by the specific application. If there is a specific computation that generalizes to many data sets I would, otherwise no. I'm not a data scientist, but my thoughts are that there is no specific computation that generalizes to many problems. Choosing the right normalization computations depends on the data you have and what you are trying to achieve.

**Comment by russorat** [comment]
thanks! seems like we might need a "statistics" package with functions for each of these normalization techniques. https://en.wikipedia.org/wiki/Normalization_(statistics)#Examples
